opm/simulators/linalg/bda/BILU0.cpp

```diff
@@ -0,0 +1,381 @@
+/*
+  Copyright 2019 Equinor ASA
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <config.h>
+#include <opm/common/OpmLog/OpmLog.hpp>
+#include <opm/common/ErrorMacros.hpp>
+
+#include <opm/simulators/linalg/bda/BdaSolver.hpp>
+#include <opm/simulators/linalg/bda/BILU0.hpp>
+#include <opm/simulators/linalg/bda/Reorder.hpp>
+
+
+namespace bda
+{
+
+    using Opm::OpmLog;
+
+    // define 'second' as 'BdaSolver<>::second', this allows usage of the second() function for timing
+    // typedefs cannot handle templates
+    const auto second = BdaSolver<>::second;
+
+    template <unsigned int block_size>
+    BILU0<block_size>::BILU0(bool level_scheduling_, bool graph_coloring_, int verbosity_) :
+        verbosity(verbosity_), level_scheduling(level_scheduling_), graph_coloring(graph_coloring_)
+    {
+        if (level_scheduling == graph_coloring) {
+            OPM_THROW(std::logic_error, "Error, either level_scheduling or graph_coloring must be true, not both\n");
+        }
+    }
+
+    template <unsigned int block_size>
+    BILU0<block_size>::~BILU0()
+    {
+        delete[] invDiagVals;
+        delete[] diagIndex;
+        delete[] rowsPerColor;
+        freeBlockedMatrix(&LUMat);
+        freeBlockedMatrix(&LMat);
+        freeBlockedMatrix(&UMat);
+        delete[] toOrder;
+        delete[] fromOrder;
+        freeBlockedMatrix(&rMat);
+    }
+
+    template <unsigned int block_size>
+    bool BILU0<block_size>::init(BlockedMatrix *mat)
+    {
+        const unsigned int bs = block_size;
+        double t1 = 0.0, t2 = 0.0;
+        BlockedMatrix *CSCmat = nullptr;
+
+        this->N = mat->Nb * block_size;
+        this->Nb = mat->Nb;
+        this->nnz = mat->nnzbs * block_size * block_size;
+        this->nnzbs = mat->nnzbs;
+
+        toOrder = new int[N];
+        fromOrder = new int[N];
```

Comment:
Both are needed in the reorderBlockedMatrixByPattern function: the toOrder array is needed to change the values of all column indices, so they go from pointing to a row in the original order to pointing to a row in the new order.
The fromOrder array is needed during the building of the rowPointers array of the reordered matrix, which needs to go through the rows of the reordered matrix sequentially starting from row 0.
It is possible to use only one order array, but it would make this function less straightforward, and potentially slower.

Reply:
Just to see where I am wrong:

In case you mean function `blocked_reorder_matrix_by_pattern` in Reorder.cpp:
Let us try to get rid of e.g. the from array. For this we need to rewrite the loop
```
for (i = 0; i < mat->Nb; i++) {
        int thisRow = fromOrder[i];
        // put thisRow from the old matrix into row i of the new matrix
        rMat->rowPointers[i + 1] = rMat->rowPointers[i] + mat->rowPointers[thisRow + 1] - mat->rowPointers[thisRow];
        for (k = mat->rowPointers[thisRow]; k < mat->rowPointers[thisRow + 1]; k++) {
            for (j = 0; j < bs * bs; j++){
                rMat->nnzValues[rIndex * bs * bs + j] = mat->nnzValues[k * bs * bs + j];
            }
            rMat->colIndices[rIndex] = mat->colIndices[k];
            rIndex++;
        }
    }
```
to
```
for (thisRow = 0; thisRow < mat->Nb; thisRow++) { // changed line
        int i = toOrder[thisRow]; // changed line, rest stays the same.
        // put thisRow from the old matrix into row i of the new matrix
        rMat->rowPointers[i + 1] = rMat->rowPointers[i] + mat->rowPointers[thisRow + 1] - mat->rowPointers[thisRow];
        for (k = mat->rowPointers[thisRow]; k < mat->rowPointers[thisRow + 1]; k++) {
            for (j = 0; j < bs * bs; j++){
                rMat->nnzValues[rIndex * bs * bs + j] = mat->nnzValues[k * bs * bs + j];
            }
            rMat->colIndices[rIndex] = mat->colIndices[k];
            rIndex++;
        }
    }
```
Is that really slower? 
If we reorder, don't we always jump around in one of the two arrays and it should not matter which? E.g. in blocked_reorder_vector_by_pattern

Reply:
Unfortunately, this does not work in this case.  rMat->rowPointers[i + 1] depends on rMat->rowPointers[i], so we cannot jump around with the i index. The way I see that the matrix reordering could use only the toOrder would be something like this:
```
int *nnzsPerRow = new int[Nb];
for (int thisRow = 0; thisRow < mat->Nb; thisRow++) {
    i = toOrder[thisRow];
    nnzsPerRow[i] = mat->rowPointers[thisRow + 1] - mat->rowPointers[thisRow];
}

for (i = 0; i < mat->Nb; i++) {
        // put thisRow from the old matrix into row i of the new matrix
        rMat->rowPointers[i + 1] = rMat->rowPointers[i] + nnzsPerRow[i]; // changed lines end here
        for (k = mat->rowPointers[thisRow]; k < mat->rowPointers[thisRow + 1]; k++) {
            for (j = 0; j < bs * bs; j++){
                rMat->nnzValues[rIndex * bs * bs + j] = mat->nnzValues[k * bs * bs + j];
            }
            rMat->colIndices[rIndex] = mat->colIndices[k];
            rIndex++;
        }
    }
```
Here, a different array, nnzsPerRow, is needed instead of the fromOrder. This array remains the same as long as the sparsity pattern does, so using it would not be slower than using toOrder (I wrote yesterday that it would be slower because I didn't consider re-using the nnzsPerRow array). It would, however, still require two arrays of size Nb to be inputs of the matrix reordering function, so I don't see a reason to use the above code over the current code.

Reply:
Got it. thanks for the clarifaction